better-words/message_parser.py

```diff
@@ -2,60 +2,70 @@
 import re
 from os import listdir
 from os.path import splitext
+from collections import namedtuple
+
+# A named tuple representing a match from the parser
+ParserMatch = namedtuple('ParserMatch', 'message category entry word')
 
 class MessageParser(object):
   def __init__(self):
-    self.words = {}
-
     # iterate over all the word categories and load up each yml file
+    words = {}
     for file in listdir('./better-words/words'):
       category, extension = splitext(file)
       if extension == '.yml':
         with open('./better-words/words/' + file, 'r') as stream:
-          self.words[category] = yaml.load(stream)
+          words[category] = yaml.load(stream)
+
+    self.words = self.parseYaml(words)
+
+  # Recursively convert a dictionary into a generic named tuple
+  # we use this to transform our YML data into an immutable tuple
+  def parseYaml(self, dictionary):
+    for key, value in dictionary.iteritems():
+      if isinstance(value, dict):
+        dictionary[key] = self.parseYaml(value)
+      if isinstance(value, list):
+        for index, v in enumerate(value):
+          if isinstance(v, dict):
```

Comment:
will it always be a dict? could it be a list? 

if we know it'll always be a dict we could do `value[index] = self.parseYaml(value)` which is less nesting and repetition